File: sam3/app/src/main/java/com/sam3/TrackerMath.kt
package com.sam3

import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Numeric primitives for the tracker host loop, ported 1:1 from
 * scripts/tracker_host_loop.py. Tensors are row-major float arrays with explicit
 * dims (NCHW like numpy). Binary masks use 64-bit packed bitsets (LSB-first bit
 * order, matching np.packbits(bitorder="little") read little-endian).
 */
internal object TM {

    fun sigmoid(x: Float): Float = (1.0 / (1.0 + exp(-x.toDouble()))).toFloat()

    // ---------------- fp16 / bf16
    fun halfToFloat(h: Int): Float {
        val s = (h ushr 15) and 0x1; val e = (h ushr 10) and 0x1F; val m = h and 0x3FF
        val bits = when {
            e == 0 -> if (m == 0) s shl 31 else {
                var mant = m; var exp2 = -1
                do { mant = mant shl 1; exp2++ } while (mant and 0x400 == 0)
                (s shl 31) or ((127 - 15 - exp2) shl 23) or ((mant and 0x3FF) shl 13)
            }
            e == 0x1F -> (s shl 31) or (0xFF shl 23) or (m shl 13)
            else -> (s shl 31) or ((e - 15 + 127) shl 23) or (m shl 13)
        }
        return Float.fromBits(bits)
    }

    /** float32 -> fp16 bits, round-to-nearest-even (like numpy .astype(float16)). */
    fun floatToHalf(f: Float): Int {
        val bits = f.toRawBits()
        val s = (bits ushr 16) and 0x8000
        val e = (bits ushr 23) and 0xFF
        var m = bits and 0x7FFFFF
        if (e == 0xFF) return s or 0x7C00 or (if (m != 0) 0x200 or (m ushr 13) else 0)
        val eH = e - 127 + 15
        if (eH >= 0x1F) return s or 0x7C00
        if (eH <= 0) {
            if (eH < -10) return s
            m = m or 0x800000
            val shift = 14 - eH
            val half = m ushr shift
            val rem = m and ((1 shl shift) - 1)
            val mid = 1 shl (shift - 1)
            var r = half
            if (rem > mid || (rem == mid && (half and 1) == 1)) r++
            return s or r
        }
        val half = m ushr 13
        val rem = m and 0x1FFF
        var r = (eH shl 10) or half
        if (rem > 0x1000 || (rem == 0x1000 && (r and 1) == 1)) r++
        return s or r
    }

    fun toHalfBits(v: Float): Short = floatToHalf(v).toShort()

    fun halfBitsToFloat(h: Short): Float = halfToFloat(h.toInt() and 0xFFFF)

    /** float32 -> bfloat16 -> float32 in place (round-to-nearest-even, like torch). */
    fun bf16InPlace(x: FloatArray) {
        for (i in x.indices) {
            val b = x[i].toRawBits()
            val r = ((b ushr 16) and 1) + 0x7FFF
            x[i] = Float.fromBits((b + r) and 0xFFFF0000.toInt())
        }
    }

    // ---------------- bilinear interpolation (align_corners=False, no antialias)
    private class AxisIdx(val i0: IntArray, val i1: IntArray, val lam: FloatArray)

    private fun axisIdx(iSz: Int, oSz: Int): AxisIdx {
        val i0 = IntArray(oSz); val i1 = IntArray(oSz); val lam = FloatArray(oSz)
        for (o in 0 until oSz) {
            var src = (o + 0.5) * (iSz.toDouble() / oSz) - 0.5
            if (src < 0.0) src = 0.0
            val f = min(floor(src).toInt(), iSz - 1)
            i0[o] = f; i1[o] = min(f + 1, iSz - 1)
            lam[o] = (src - f).toFloat()
        }
        return AxisIdx(i0, i1, lam)
    }

    /**
     * torch F.interpolate(mode="bilinear", align_corners=False), planes-major input
     * (n*c, ih, iw) -> (n*c, oh, ow). Rows first, then columns, f32 arithmetic —
     * the same evaluation order as the numpy reference.
     */
    fun interpBilinear(x: FloatArray, planes: Int, ih: Int, iw: Int, oh: Int, ow: Int): FloatArray {
        if (ih == oh && iw == ow) return x.copyOf()
        val ay = axisIdx(ih, oh)
        val ax = axisIdx(iw, ow)
        val out = FloatArray(planes * oh * ow)
        val top = FloatArray(iw)
        for (p in 0 until planes) {
            val base = p * ih * iw
            val obase = p * oh * ow
            for (y in 0 until oh) {
                val r0 = base + ay.i0[y] * iw
                val r1 = base + ay.i1[y] * iw
                val ly = ay.lam[y]
                for (xw in 0 until iw) top[xw] = x[r0 + xw] * (1f - ly) + x[r1 + xw] * ly
                val orow = obase + y * ow
                for (xo in 0 until ow) {
                    val lx = ax.lam[xo]
                    out[orow + xo] = top[ax.i0[xo]] * (1f - lx) + top[ax.i1[xo]] * lx
                }
            }
        }
        return out
    }

    // ---------------- antialiased bilinear (triangle filter, PIL-style)
    private class AAWeights(val xmin: IntArray, val xmax: IntArray, val ws: Array<FloatArray>)

    private val aaCache = HashMap<Long, AAWeights>()

    private fun aaWeights(iSz: Int, oSz: Int): AAWeights {
        val key = iSz.toLong() shl 32 or oSz.toLong()
        aaCache[key]?.let { return it }
        val scale = iSz.toDouble() / oSz
        val support = max(scale, 1.0)
        val xmin = IntArray(oSz); val xmax = IntArray(oSz)
        val ws = Array(oSz) { FloatArray(0) }
        for (o in 0 until oSz) {
            val center = (o + 0.5) * scale
            val lo = max(0, (center - support + 0.5).toInt())
            val hi = min(iSz, (center + support + 0.5).toInt())
            xmin[o] = lo; xmax[o] = hi
            val w = FloatArray(hi - lo)
            var s = 0.0
            for (k in lo until hi) {
                val v = max(1.0 - abs((k + 0.5 - center) / support), 0.0)
                w[k - lo] = v.toFloat(); s += v
            }
            if (s > 0) for (i in w.indices) w[i] = (w[i] / s).toFloat()
            ws[o] = w
        }
        val r = AAWeights(xmin, xmax, ws)
        aaCache[key] = r
        return r
    }

    /** torch F.interpolate(..., antialias=True), planes-major (n*c, ih, iw). */
    fun interpBilinearAA(x0: FloatArray, planes: Int, ih: Int, iw: Int, oh: Int, ow: Int): FloatArray {
        var x = x0
        var h = ih
        if (ih != oh) {
            val a = aaWeights(ih, oh)
            val out = FloatArray(planes * oh * iw)
            for (p in 0 until planes) {
                val base = p * ih * iw; val obase = p * oh * iw
                for (o in 0 until oh) {
                    val w = a.ws[o]; val lo = a.xmin[o]
                    val orow = obase + o * iw
                    for (k in w.indices) {
                        val wk = w[k]; val row = base + (lo + k) * iw
                        for (c in 0 until iw) out[orow + c] += wk * x[row + c]
                    }
                }
            }
            x = out; h = oh
        }
        if (iw != ow) {
            val a = aaWeights(iw, ow)
            val out = FloatArray(planes * h * ow)
            for (p in 0 until planes) {
                val base = p * h * iw; val obase = p * h * ow
                for (r in 0 until h) {
                    val row = base + r * iw; val orow = obase + r * ow
                    for (o in 0 until ow) {
                        val w = a.ws[o]; val lo = a.xmin[o]
                        var acc = 0f
                        for (k in w.indices) acc += w[k] * x[row + lo + k]
                        out[orow + o] = acc
                    }
                }
            }
            x = out
        }
        return x
    }

    // ---------------- erf / GELU
    /** erf via the Numerical Recipes erfc rational approximation (|err| < 1.2e-7). */
    fun erf(x: Double): Double {
        val z = abs(x)
        val t = 1.0 / (1.0 + 0.5 * z)
        val ans = t * exp(-z * z - 1.26551223 + t * (1.00002368 + t * (0.37409196 +
            t * (0.09678418 + t * (-0.18628806 + t * (0.27886807 + t * (-1.13520398 +
            t * (1.48851587 + t * (-0.82215223 + t * 0.17087277)))))))))
        val erfc = if (x >= 0) ans else 2.0 - ans
        return 1.0 - erfc
    }

    fun geluInPlace(x: FloatArray) {
        for (i in x.indices) {
            val v = x[i].toDouble()
            x[i] = (0.5 * v * (1.0 + erf(v / sqrt(2.0)))).toFloat()
        }
    }

    // ---------------- small NN ops (host-side conv chain, LN, linears)
    /** Conv2d with stride == kernel (non-overlapping blocks), NCHW. */
    fun convBlock(x: FloatArray, n: Int, ci: Int, h: Int, w: Int,
                  wgt: FloatArray, co: Int, k: Int, bias: FloatArray): FloatArray {
        val oh = h / k; val ow = w / k
        val out = FloatArray(n * co * oh * ow)
        for (b in 0 until n) {
            for (o in 0 until co) {
                val obase = (b * co + o) * oh * ow
                for (y in 0 until oh) for (xw in 0 until ow) {
                    var acc = 0f
                    for (c in 0 until ci) {
                        val ibase = (b * ci + c) * h * w + y * k * w + xw * k
                        val wbase = (o * ci + c) * k * k
                        for (ky in 0 until k) for (kx in 0 until k) {
                            acc += x[ibase + ky * w + kx] * wgt[wbase + ky * k + kx]
                        }
                    }
                    out[obase + y * ow + xw] = acc + bias[o]
                }
            }
        }
        return out
    }

    /** LayerNorm over the channel axis of NCHW (eps 1e-6), in place. */
    fun layerNorm2dInPlace(x: FloatArray, n: Int, c: Int, h: Int, w: Int,
                           wgt: FloatArray, bias: FloatArray) {
        val hw = h * w
        for (b in 0 until n) for (p in 0 until hw) {
            var mu = 0f
            for (ch in 0 until c) mu += x[(b * c + ch) * hw + p]
            mu /= c
            var va = 0f
            for (ch in 0 until c) { val d = x[(b * c + ch) * hw + p] - mu; va += d * d }
            va /= c
            val inv = (1.0 / sqrt(va.toDouble() + 1e-6)).toFloat()
            for (ch in 0 until c) {
                val i = (b * c + ch) * hw + p
                x[i] = (x[i] - mu) * inv * wgt[ch] + bias[ch]
            }
        }
    }

    /** x (m,inDim) @ w(out,inDim)^T + b -> (m,out). */
    fun linear(x: FloatArray, m: Int, inDim: Int, w: FloatArray, out: Int, b: FloatArray): FloatArray {
        val y = FloatArray(m * out)
        for (r in 0 until m) {
            val xb = r * inDim
            for (o in 0 until out) {
                var acc = 0f
                val wb = o * inDim
                for (i in 0 until inDim) acc += x[xb + i] * w[wb + i]
                y[r * out + o] = acc + b[o]
            }
        }
        return y
    }

    /** Sine PE of get_1d_sine_pe: pos (n) -> (n, dim), [sin | cos]. */
    fun sine1dPe(pos: FloatArray, dim: Int): FloatArray {
        val peDim = dim / 2
        val out = FloatArray(pos.size * dim)
        for (r in pos.indices) {
            for (i in 0 until peDim) {
                val dimT = Math.pow(10000.0, 2.0 * (i / 2) / peDim)
                val v = pos[r] / dimT
                out[r * dim + i] = Math.sin(v).toFloat()
                out[r * dim + peDim + i] = Math.cos(v).toFloat()
            }
        }
        return out
    }

    // ---------------- packed binary masks
    fun packedWords(px: Int): Int = (px + 63) ushr 6

    /** Pack (src[offset + i] > thresh) for i in 0 until px, LSB-first. */
    fun pack(src: FloatArray, offset: Int, px: Int, thresh: Float): LongArray {
        val w = LongArray(packedWords(px))
        for (i in 0 until px) {
            if (src[offset + i] > thresh) w[i ushr 6] = w[i ushr 6] or (1L shl (i and 63))
        }
        return w
    }

    fun popcount(a: LongArray): Int {
        var s = 0
        for (w in a) s += java.lang.Long.bitCount(w)
        return s
    }

    fun popcountAnd(a: LongArray, b: LongArray): Int {
        var s = 0
        for (i in a.indices) s += java.lang.Long.bitCount(a[i] and b[i])
        return s
    }

    fun popcountOr(a: LongArray, b: LongArray): Int {
        var s = 0
        for (i in a.indices) s += java.lang.Long.bitCount(a[i] or b[i])
        return s
    }

    fun testBit(a: LongArray, i: Int): Boolean = (a[i ushr 6] ushr (i and 63)) and 1L == 1L
}

File: sam3/ios/Sources/TrackerMath.swift
import Foundation

/// Numeric primitives for the tracker host loop, ported 1:1 from the Kotlin port
/// (app/src/main/java/com/sam3/TrackerMath.kt), itself ported from
/// scripts/tracker_host_loop.py. Tensors are row-major float arrays with explicit
/// dims (NCHW like numpy). Binary masks use 64-bit packed bitsets (LSB-first bit
/// order, matching np.packbits(bitorder="little") read little-endian).
enum TM {
    static func sigmoid(_ x: Float) -> Float {
        Float(1.0 / (1.0 + exp(-Double(x))))
    }

    // ---------------- fp16 / bf16 (arm64 Float16 rounds RNE, same as numpy)
    static func toHalfBits(_ v: Float) -> UInt16 { Float16(v).bitPattern }

    static func halfBitsToFloat(_ h: UInt16) -> Float { Float(Float16(bitPattern: h)) }

    /// float32 -> bfloat16 -> float32 in place (round-to-nearest-even, like torch).
    static func bf16InPlace(_ x: inout [Float]) {
        x.withUnsafeMutableBufferPointer { p in
            for i in 0..<p.count {
                let b = p[i].bitPattern
                let r = ((b >> 16) & 1) &+ 0x7FFF
                p[i] = Float(bitPattern: (b &+ r) & 0xFFFF_0000)
            }
        }
    }

    // ---------------- bilinear interpolation (align_corners=False, no antialias)
    private struct AxisIdx {
        let i0: [Int]
        let i1: [Int]
        let lam: [Float]
    }

    private static func axisIdx(_ iSz: Int, _ oSz: Int) -> AxisIdx {
        var i0 = [Int](repeating: 0, count: oSz)
        var i1 = [Int](repeating: 0, count: oSz)
        var lam = [Float](repeating: 0, count: oSz)
        for o in 0..<oSz {
            var src = (Double(o) + 0.5) * (Double(iSz) / Double(oSz)) - 0.5
            if src < 0 { src = 0 }
            let f = min(Int(src.rounded(.down)), iSz - 1)
            i0[o] = f
            i1[o] = min(f + 1, iSz - 1)
            lam[o] = Float(src - Double(f))
        }
        return AxisIdx(i0: i0, i1: i1, lam: lam)
    }

    /// torch F.interpolate(mode="bilinear", align_corners=False), planes-major input
    /// (n*c, ih, iw) -> (n*c, oh, ow). Rows first, then columns, f32 arithmetic —
    /// the same evaluation order as the numpy reference.
    static func interpBilinear(_ x: [Float], _ planes: Int, _ ih: Int, _ iw: Int,
                               _ oh: Int, _ ow: Int) -> [Float] {
        if ih == oh && iw == ow { return x }
        let ay = axisIdx(ih, oh)
        let ax = axisIdx(iw, ow)
        var out = [Float](repeating: 0, count: planes * oh * ow)
        var top = [Float](repeating: 0, count: iw)
        x.withUnsafeBufferPointer { xp in
            out.withUnsafeMutableBufferPointer { op in
                top.withUnsafeMutableBufferPointer { tp in
                    for p in 0..<planes {
                        let base = p * ih * iw
                        let obase = p * oh * ow
                        for y in 0..<oh {
                            let r0 = base + ay.i0[y] * iw
                            let r1 = base + ay.i1[y] * iw
                            let ly = ay.lam[y]
                            for xw in 0..<iw { tp[xw] = xp[r0 + xw] * (1 - ly) + xp[r1 + xw] * ly }
                            let orow = obase + y * ow
                            for xo in 0..<ow {
                                let lx = ax.lam[xo]
                                op[orow + xo] = tp[ax.i0[xo]] * (1 - lx) + tp[ax.i1[xo]] * lx
                            }
                        }
                    }
                }
            }
        }
        return out
    }

    // ---------------- antialiased bilinear (triangle filter, PIL-style)
    private struct AAWeights {
        let xmin: [Int]
        let ws: [[Float]]
    }

    private static var aaCache = [Int64: AAWeights]()

    private static func aaWeights(_ iSz: Int, _ oSz: Int) -> AAWeights {
        let key = Int64(iSz) << 32 | Int64(oSz)
        if let c = aaCache[key] { return c }
        let scale = Double(iSz) / Double(oSz)
        let support = max(scale, 1.0)
        var xmin = [Int](repeating: 0, count: oSz)
        var ws = [[Float]]()
        ws.reserveCapacity(oSz)
        for o in 0..<oSz {
            let center = (Double(o) + 0.5) * scale
            let lo = max(0, Int(center - support + 0.5))
            let hi = min(iSz, Int(center + support + 0.5))
            xmin[o] = lo
            var w = [Float](repeating: 0, count: hi - lo)
            var s = 0.0
            for k in lo..<hi {
                let v = max(1.0 - abs((Double(k) + 0.5 - center) / support), 0.0)
                w[k - lo] = Float(v)
                s += v
            }
            if s > 0 { for i in 0..<w.count { w[i] = Float(Double(w[i]) / s) } }
            ws.append(w)
        }
        let r = AAWeights(xmin: xmin, ws: ws)
        aaCache[key] = r
        return r
    }

    /// torch F.interpolate(..., antialias=True), planes-major (n*c, ih, iw).
    static func interpBilinearAA(_ x0: [Float], _ planes: Int, _ ih: Int, _ iw: Int,
                                 _ oh: Int, _ ow: Int) -> [Float] {
        var x = x0
        var h = ih
        if ih != oh {
            let a = aaWeights(ih, oh)
            var out = [Float](repeating: 0, count: planes * oh * iw)
            x.withUnsafeBufferPointer { xp in
                out.withUnsafeMutableBufferPointer { op in
                    for p in 0..<planes {
                        let base = p * ih * iw
                        let obase = p * oh * iw
                        for o in 0..<oh {
                            let w = a.ws[o]
                            let lo = a.xmin[o]
                            let orow = obase + o * iw
                            for k in 0..<w.count {
                                let wk = w[k]
                                let row = base + (lo + k) * iw
                                for c in 0..<iw { op[orow + c] += wk * xp[row + c] }
                            }
                        }
                    }
                }
            }
            x = out
            h = oh
        }
        if iw != ow {
            let a = aaWeights(iw, ow)
            var out = [Float](repeating: 0, count: planes * h * ow)
            x.withUnsafeBufferPointer { xp in
                out.withUnsafeMutableBufferPointer { op in
                    for p in 0..<planes {
                        let base = p * h * iw
                        let obase = p * h * ow
                        for r in 0..<h {
                            let row = base + r * iw
                            let orow = obase + r * ow
                            for o in 0..<ow {
                                let w = a.ws[o]
                                let lo = a.xmin[o]
                                var acc: Float = 0
                                for k in 0..<w.count { acc += w[k] * xp[row + lo + k] }
                                op[orow + o] = acc
                            }
                        }
                    }
                }
            }
            x = out
        }
        return x
    }

    // ---------------- GELU (libm erf: correctly rounded, matches numpy)
    static func geluInPlace(_ x: inout [Float]) {
        for i in 0..<x.count {
            let v = Double(x[i])
            x[i] = Float(0.5 * v * (1.0 + erf(v / 2.0.squareRoot())))
        }
    }

    // ---------------- small NN ops (host-side conv chain, LN, linears)
    /// Conv2d with stride == kernel (non-overlapping blocks), NCHW.
    static func convBlock(_ x: [Float], _ n: Int, _ ci: Int, _ h: Int, _ w: Int,
                          _ wgt: [Float], _ co: Int, _ k: Int, _ bias: [Float]) -> [Float] {
        let oh = h / k
        let ow = w / k
        var out = [Float](repeating: 0, count: n * co * oh * ow)
        x.withUnsafeBufferPointer { xp in
            wgt.withUnsafeBufferPointer { wp in
                out.withUnsafeMutableBufferPointer { op in
                    for b in 0..<n {
                        for o in 0..<co {
                            let obase = (b * co + o) * oh * ow
                            for y in 0..<oh {
                                for xw in 0..<ow {
                                    var acc: Float = 0
                                    for c in 0..<ci {
                                        let ibase = (b * ci + c) * h * w + y * k * w + xw * k
                                        let wbase = (o * ci + c) * k * k
                                        for ky in 0..<k {
                                            for kx in 0..<k {
                                                acc += xp[ibase + ky * w + kx] * wp[wbase + ky * k + kx]
                                            }
                                        }
                                    }
                                    op[obase + y * ow + xw] = acc + bias[o]
                                }
                            }
                        }
                    }
                }
            }
        }
        return out
    }

    /// LayerNorm over the channel axis of NCHW (eps 1e-6), in place.
    static func layerNorm2dInPlace(_ x: inout [Float], _ n: Int, _ c: Int, _ h: Int, _ w: Int,
                                   _ wgt: [Float], _ bias: [Float]) {
        let hw = h * w
        x.withUnsafeMutableBufferPointer { p in
            for b in 0..<n {
                for px in 0..<hw {
                    var mu: Float = 0
                    for ch in 0..<c { mu += p[(b * c + ch) * hw + px] }
                    mu /= Float(c)
                    var va: Float = 0
                    for ch in 0..<c {
                        let d = p[(b * c + ch) * hw + px] - mu
                        va += d * d
                    }
                    va /= Float(c)
                    let inv = Float(1.0 / (Double(va) + 1e-6).squareRoot())
                    for ch in 0..<c {
                        let i = (b * c + ch) * hw + px
                        p[i] = (p[i] - mu) * inv * wgt[ch] + bias[ch]
                    }
                }
            }
        }
    }

    /// x (m,inDim) @ w(out,inDim)^T + b -> (m,out).
    static func linear(_ x: [Float], _ m: Int, _ inDim: Int,
                       _ w: [Float], _ out: Int, _ b: [Float]) -> [Float] {
        var y = [Float](repeating: 0, count: m * out)
        x.withUnsafeBufferPointer { xp in
            w.withUnsafeBufferPointer { wp in
                y.withUnsafeMutableBufferPointer { yp in
                    for r in 0..<m {
                        let xb = r * inDim
                        for o in 0..<out {
                            var acc: Float = 0
                            let wb = o * inDim
                            for i in 0..<inDim { acc += xp[xb + i] * wp[wb + i] }
                            yp[r * out + o] = acc + b[o]
                        }
                    }
                }
            }
        }
        return y
    }

    /// Sine PE of get_1d_sine_pe: pos (n) -> (n, dim), [sin | cos].
    static func sine1dPe(_ pos: [Float], _ dim: Int) -> [Float] {
        let peDim = dim / 2
        var out = [Float](repeating: 0, count: pos.count * dim)
        for r in 0..<pos.count {
            for i in 0..<peDim {
                let dimT = pow(10000.0, 2.0 * Double(i / 2) / Double(peDim))
                let v = Double(pos[r]) / dimT
                out[r * dim + i] = Float(sin(v))
                out[r * dim + peDim + i] = Float(cos(v))
            }
        }
        return out
    }

    // ---------------- packed binary masks
    static func packedWords(_ px: Int) -> Int { (px + 63) >> 6 }

    /// Pack (src[offset + i] > thresh) for i in 0..<px, LSB-first.
    static func pack(_ src: [Float], _ offset: Int, _ px: Int, _ thresh: Float) -> [UInt64] {
        var w = [UInt64](repeating: 0, count: packedWords(px))
        src.withUnsafeBufferPointer { sp in
            w.withUnsafeMutableBufferPointer { wp in
                for i in 0..<px {
                    if sp[offset + i] > thresh { wp[i >> 6] |= 1 << UInt64(i & 63) }
                }
            }
        }
        return w
    }

    static func popcount(_ a: [UInt64]) -> Int {
        var s = 0
        for w in a { s += w.nonzeroBitCount }
        return s
    }

    static func popcountAnd(_ a: [UInt64], _ b: [UInt64]) -> Int {
        var s = 0
        for i in 0..<a.count { s += (a[i] & b[i]).nonzeroBitCount }
        return s
    }

    static func popcountOr(_ a: [UInt64], _ b: [UInt64]) -> Int {
        var s = 0
        for i in 0..<a.count { s += (a[i] | b[i]).nonzeroBitCount }
        return s
    }

    static func testBit(_ a: [UInt64], _ i: Int) -> Bool {
        (a[i >> 6] >> UInt64(i & 63)) & 1 == 1
    }
}
